lib/auth/methods.go

```diff
@@ -186,37 +213,62 @@ func (s *Server) authenticateUser(ctx context.Context, req AuthenticateUserReque
 		// registered.
 		devs, err := s.Identity.GetMFADevices(ctx, user, false /* withSecrets */)
 		if err != nil && !trace.IsNotFound(err) {
-			return nil, trace.Wrap(err)
+			return nil, "", trace.Wrap(err)
 		}
 		if len(devs) != 0 {
 			log.Warningf("MFA bypass attempt by user %q, access denied.", user)
-			return nil, trace.AccessDenied("missing second factor authentication")
+			return nil, "", trace.AccessDenied("missing second factor authentication")
 		}
 	default:
 		// Some form of MFA is required but none provided. Either client is
 		// buggy (didn't send MFA response) or someone is trying to bypass
 		// MFA.
 		log.Warningf("MFA bypass attempt by user %q, access denied.", user)
-		return nil, trace.AccessDenied("missing second factor")
+		return nil, "", trace.AccessDenied("missing second factor")
 	}
 	if err = s.WithUserLock(user, func() error {
 		return s.checkPasswordWOToken(user, req.Pass.Password)
 	}); err != nil {
 		if fieldErr := getErrorByTraceField(err); fieldErr != nil {
-			return nil, trace.Wrap(fieldErr)
+			return nil, "", trace.Wrap(fieldErr)
 		}
 		// provide obscure message on purpose, while logging the real
 		// error server side
 		log.Debugf("User %v failed to authenticate: %v.", user, err)
-		return nil, trace.AccessDenied("invalid username or password")
+		return nil, "", trace.AccessDenied("invalid username or password")
 	}
-	return nil, nil
+	return nil, user, nil
+}
+
+func (s *Server) authenticatePasswordless(ctx context.Context, req AuthenticateUserRequest) (*types.MFADevice, string, error) {
+	mfaResponse := &proto.MFAAuthenticateResponse{
+		Response: &proto.MFAAuthenticateResponse_Webauthn{
+			Webauthn: wanlib.CredentialAssertionResponseToProto(req.Webauthn),
+		},
+	}
+	dev, user, err := s.validateMFAAuthResponse(ctx, mfaResponse, "", true /* passwordless */)
+	if err != nil {
+		log.Debugf("Passwordless authentication failed: %v", err)
```

Comment:
Should this be logged at a Warn or Err level since it seems to be indicating an error? Same for the log entry below.

Reply:
There are valid reasons for authentication to fail, it doesn't necessarily mean a noticeable event (or bug). This is akin to the [debug statement on authenticateUser](https://github.com/gravitational/teleport/blob/61fce15ee218a29c425ba4082b37e81ab0dcdec3/lib/auth/methods.go#L153).

Kept.